capsules/extra/src/eeprom.rs

```diff
@@ -0,0 +1,259 @@
+//! > interface.
+//!
+//! Usage
+//! -----
+//!
+//! ```rust
+//! let i2cmux = I2CMuxComponent::new(i2c0, None).finalize(components::i2c_mux_component_static!());
+//!
+//! let eeprom_buffer = static_init!([u8; 34], [0; 34]);
+//!
+//! let eeprom_i2c_device = static_init!(I2CDevice, I2CDevice::new(i2cmux, 0x50));
+//! let eeprom_capsule = static_init!(capsules_extra::eeprom::EEPROM,capsules_extra::eeprom::EEPROM::new(
+//!             eeprom_i2c_device,
+//!             eeprom_buffer,
+//!         ) );
+//! eeprom_i2c_device.set_client(eeprom_capsule);
+//!
+//! let nonvolatile_storage = components::nonvolatile_storage::NonvolatileStorageComponent::new(
+//!         board_kernel,
+//!         capsules_extra::nonvolatile_storage_driver::DRIVER_NUM,
+//!         eeprom_capsule,
+//!         0x0,
+//!         0x10000,
+//!         0x0,
+//!         0x0,
+//!     ).finalize(components::nonvolatile_storage_component_static!(capsules_extra::eeprom::EEPROM));
+//! ```
+
+use core::cell::Cell;
+use core::cmp;
+
+use kernel::hil::i2c::{Error, I2CClient, I2CDevice};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::{hil, ErrorCode};
+
+const PAGE_SIZE: usize = 32;
+
+pub struct EEPROMPage(pub [u8; PAGE_SIZE]);
+
+impl Default for EEPROMPage {
+    fn default() -> Self {
+        Self([0; PAGE_SIZE])
+    }
+}
+
+impl AsMut<[u8]> for EEPROMPage {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum State {
+    Idle,
+    Reading,
+    Writing,
+    Erasing,
+}
+
+pub struct EEPROM<'a> {
+    i2c: &'a dyn I2CDevice,
+    buffer: TakeCell<'static, [u8]>,
+    client_page: TakeCell<'a, EEPROMPage>,
+    flash_client: OptionalCell<&'a dyn hil::flash::Client<EEPROM<'a>>>,
+    state: Cell<State>,
+}
+
+impl<'a> EEPROM<'a> {
+    pub fn new(i2c: &'a dyn I2CDevice, buffer: &'static mut [u8]) -> Self {
+        Self {
+            i2c,
+            buffer: TakeCell::new(buffer),
+            client_page: TakeCell::empty(),
+            flash_client: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+        }
+    }
+
+    fn read_sector(
+        &self,
+        page_number: usize,
+        buf: &'static mut EEPROMPage,
+    ) -> Result<(), (ErrorCode, &'static mut EEPROMPage)> {
+        self.i2c.enable();
+        let address = page_number * PAGE_SIZE;
+        if let Some(rxbuffer) = self.buffer.take() {
+            rxbuffer[0] = ((address >> 8) & 0x00ff) as u8;
+            rxbuffer[1] = (address & 0x00ff) as u8;
+
+            self.state.set(State::Reading);
+            if let Err((error, local_buffer)) = self.i2c.write_read(rxbuffer, 2, PAGE_SIZE) {
+                self.buffer.replace(local_buffer);
+                self.i2c.disable();
+                Err((error.into(), buf))
+            } else {
+                self.client_page.replace(buf);
+                Ok(())
+            }
+        } else {
+            Err((ErrorCode::RESERVE, buf))
+        }
```

Comment:
Shouldn't this branch also disable the `i2c`? Or is the absence of a buffer to `take()` a sign that a transaction is in progress?

A separate solution, which might be better, is to only enable after taking the buffer.

This applies to `read_sector()`, `write_sector()`, and `erase_sector()`.

Reply:
tbh this should be inside the `if`, after taking the buffer, moved